### PR TITLE
added test using gmsh python API

### PR DESCRIPTION
**Pull request by j-zimmermann**
Uses the gmsh sdk, which provides a python API.
I reimplemented the pygmsh test; pytest passed for single thread case.

Still due: thorough testing of MPI functionality.

### CODE REVIEW

**Comment by j-zimmermann** [comment]
The reason why all checks fail is that the Dockerfile needs to be updated to add the gmsh python API to the PYTHONPATH. The last commit (changing the config.yml file) was a try to circumvent this issue, but it didn't work.

File: python/test/unit/mesh/test_mesh.py [hunk 1]
```diff
@@ -205,6 +205,104 @@ def test_mesh_construction_pygmsh():
     assert mesh.topology.dim == 2
 
 
+def test_mesh_construction_gmsh():
+    import gmsh
+
+    # map from gmsh to type that is also used by meshio
```
Comment (michalhabera) [inline python/test/unit/mesh/test_mesh.py]: Please make first word capitalised in comments, we follow pep8 standard, https://www.python.org/dev/peps/pep-0008/#comments

File: python/test/unit/mesh/test_mesh.py [hunk 2]
```diff
@@ -205,6 +205,104 @@ def test_mesh_construction_pygmsh():
...
+    gmsh.initialize(sys.argv)
+    gmsh.model.add('ball')
+    gmsh.model.occ.addSphere(0.0, 0.0, 0.0, 1.0)
+    gmsh.model.occ.synchronize()
+    gmsh.model.mesh.generate()
+    nodeTags, coord, parametricCoord = gmsh.model.mesh.getNodes()
+    dim = 3
+    # reshape to get right format
+    points = numpy.reshape(coord, (numpy.int(coord.size / dim), dim))
+    print(points)
```
Comment (michalhabera) [inline python/test/unit/mesh/test_mesh.py]: Remove print() statements

File: python/test/unit/mesh/test_mesh.py [hunk 3]
```diff
@@ -205,6 +205,102 @@ def test_mesh_construction_pygmsh():
...
+    gmsh.initialize(sys.argv)
+    gmsh.model.add('ball')
+    gmsh.model.occ.addSphere(0.0, 0.0, 0.0, 1.0)
+    gmsh.model.occ.synchronize()
+    gmsh.model.mesh.generate()
+    nodeTags, coord, parametricCoord = gmsh.model.mesh.getNodes()
+    dim = 3
+    # Reshape to get right format
+    points = numpy.reshape(coord, (numpy.int(coord.size / dim), dim))
+    elementTypes, elementTags, nodeTags = gmsh.model.mesh.getElements()
```
Comment (garth-wells) [inline python/test/unit/mesh/test_mesh.py]: We avoid using `camelCaps` in DOLFIN. Use `foo_bar` instead.

File: python/test/unit/mesh/test_mesh.py [hunk 4]
```diff
@@ -205,6 +205,102 @@ def test_mesh_construction_pygmsh():
...
+    # Generate mesh
+    gmsh.initialize(sys.argv)
+    gmsh.model.add('ball')
+    gmsh.model.occ.addSphere(0.0, 0.0, 0.0, 1.0)
+    gmsh.model.occ.synchronize()
+    gmsh.model.mesh.generate()
+    nodeTags, coord, parametricCoord = gmsh.model.mesh.getNodes()
+    dim = 3
+    # Reshape to get right format
+    points = numpy.reshape(coord, (numpy.int(coord.size / dim), dim))
```
Comment (garth-wells) [inline python/test/unit/mesh/test_mesh.py]: Is `numpy.int` required? Would `int` suffice?

File: python/test/unit/mesh/test_mesh.py [hunk 5]
```diff
@@ -205,6 +205,102 @@ def test_mesh_construction_pygmsh():
...
+             }
+    # Generate mesh
+    gmsh.initialize(sys.argv)
+    gmsh.model.add('ball')
+    gmsh.model.occ.addSphere(0.0, 0.0, 0.0, 1.0)
+    gmsh.model.occ.synchronize()
+    gmsh.model.mesh.generate()
+    nodeTags, coord, parametricCoord = gmsh.model.mesh.getNodes()
+    dim = 3
+    # Reshape to get right format
```
Comment (garth-wells) [inline python/test/unit/mesh/test_mesh.py]: We generally have a blank line before comments.

File: python/test/unit/mesh/test_mesh.py [hunk 6]
```diff
@@ -205,6 +205,102 @@ def test_mesh_construction_pygmsh():
...
+    points = numpy.reshape(coord, (numpy.int(coord.size / dim), dim))
+    elementTypes, elementTags, nodeTags = gmsh.model.mesh.getElements()
+    cells = {}
```
Comment (garth-wells) [inline python/test/unit/mesh/test_mesh.py]: Move below comment.

File: python/test/unit/mesh/test_mesh.py [hunk 7]
```diff
@@ -205,6 +205,102 @@ def test_mesh_construction_pygmsh():
...
+    # Generate cells dict
+    for num, element in enumerate(elementTypes):
```
Comment (garth-wells) [inline python/test/unit/mesh/test_mesh.py]: Could this be done more efficiently, i.e. without a plain Python loop?

File: python/test/unit/mesh/test_mesh.py [hunk 8]
```diff
@@ -205,6 +205,102 @@ def test_mesh_construction_pygmsh():
...
+        name = mapping[element]
+        dim = nodes[name]
+        cells[name] = nodeTags[num] - 1  # since nodes are numbered starting from 0
+        cells[name] = numpy.reshape(cells[name], (numpy.int(cells[name].size / dim), dim))
+    # End Generate mesh
+    gmsh.finalize()
+
+    mesh = dolfin.cpp.mesh.Mesh(dolfin.MPI.comm_world, dolfin.cpp.mesh.CellType.Type.tetrahedron, points,
```
Comment (garth-wells) [inline python/test/unit/mesh/test_mesh.py]: What happens in parallel?

File: python/test/unit/mesh/test_mesh.py [hunk 9]
```diff
@@ -205,6 +205,102 @@ def test_mesh_construction_pygmsh():
...
+    # Generate 2nd order mesh
+    gmsh.model.add('ball')
+    gmsh.model.occ.addSphere(0.0, 0.0, 0.0, 1.0)
+    gmsh.model.occ.synchronize()
+    gmsh.model.mesh.generate()
+    gmsh.model.mesh.setOrder(2)
+    nodeTags, coord, parametricCoord = gmsh.model.mesh.getNodes()
+    dim = 3
+    # Reshape to get right format
+    points = numpy.reshape(coord, (numpy.int(coord.size / dim), dim))
```
Comment (garth-wells) [inline python/test/unit/mesh/test_mesh.py]: Could code form the first part of the test be re-used? Maybe use pytest  test parameterisation.

**Comment by jorgensd** [comment]
This has to be updated due to [change in dolfin mesh input format](https://github.com/FEniCS/dolfinx/pull/604)

**Comment by jorgensd** [comment]
This PR is now heavily outdated, and has been preceded by #1019 and several commits after this. @garth-wells should we close this PR?

**Pull request closed without merging**